### PR TITLE
Use elaboration to turn ~const bounds into non-const bounds

### PR DESCRIPTION
**Pull request by compiler-errors**
Feels slightly less ad-hoc, and is an attempt at this comment: https://github.com/rust-lang/rust/pull/116756#issuecomment-1763943843

The only thing that I don't like is the filtering in `one_bound_for_assoc_type` -- I wonder if I should filter somewhere else?

r? @fee1-dead or @oli-obk

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/fn_ctxt/mod.rs [hunk 1]
```diff
@@ -221,14 +221,14 @@ impl<'a, 'tcx> AstConv<'tcx> for FnCtxt<'a, 'tcx> {
...
-        // HACK(eddyb) should get the original `Span`.
-        let span = tcx.def_span(def_id);
         ty::GenericPredicates {
             parent: None,
             predicates: tcx.arena.alloc_from_iter(
                 self.param_env.caller_bounds().iter().filter_map(|predicate| {
                     match predicate.kind().skip_binder() {
                         ty::ClauseKind::Trait(data) if data.self_ty().is_param(index) => {
+                            // HACK(eddyb) should get the original `Span`.
+                            let span = tcx.def_span(def_id);
```
Comment (fee1-dead) [inline compiler/rustc_hir_typeck/src/fn_ctxt/mod.rs:231]: Should this be reverted?

**Comment by oli-obk** [comment]
> The only thing that I don't like is the filtering in `one_bound_for_assoc_type` -- I wonder if I should filter somewhere else?

😅 this is the same thing I had in my own attempt at effect generics last year. I'll compare the impls to see if I found some other subtle issues due to the higher test coverage compared to the current effects impl

**Comment by compiler-errors** [comment]
This is superseded (in the other direction) by #119721.

**Pull request closed without merging**